x/cronos/keeper/evm_log_handlers.go

```diff
@@ -0,0 +1,154 @@
+package keeper
+
+import (
+	"errors"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	gravitytypes "github.com/peggyjv/gravity-bridge/module/x/gravity/types"
+
+	"github.com/crypto-org-chain/cronos/x/cronos/types"
+)
+
+var (
+	_ types.EvmLogHandler = NativeTransferHandler{}
+	_ types.EvmLogHandler = EthereumTransferHandler{}
+)
+
+var (
+	// NativeTransferEvent represent the signature of
+	// `event __CronosNativeTransfer(address recipient, uint256 amount, string denom)`
+	NativeTransferEvent abi.Event
+
+	// EthereumTransferEvent represent the signature of
+	// `event __CronosEthereumTransfer(address recipient, uint256 amount, uint256 bridge_fee)`
+	EthereumTransferEvent abi.Event
+)
+
+func init() {
+	addressType, _ := abi.NewType("address", "", nil)
+	uint256Type, _ := abi.NewType("uint256", "", nil)
+	NativeTransferEvent = abi.NewEvent(
+		"__CronosNativeTransfer",
+		"__CronosNativeTransfer",
+		false,
+		abi.Arguments{abi.Argument{
+			Name:    "recipient",
+			Type:    addressType,
+			Indexed: false,
+		}, abi.Argument{
+			Name:    "amount",
+			Type:    uint256Type,
+			Indexed: false,
+		}},
+	)
+	EthereumTransferEvent = abi.NewEvent(
+		"__CronosEthereumTransfer",
+		"__CronosEthereumTransfer",
+		false,
+		abi.Arguments{abi.Argument{
+			Name:    "recipient",
+			Type:    addressType,
+			Indexed: false,
+		}, abi.Argument{
+			Name:    "amount",
+			Type:    uint256Type,
+			Indexed: false,
+		}, abi.Argument{
+			Name:    "bridge_fee",
+			Type:    uint256Type,
+			Indexed: false,
+		}},
+	)
+}
+
+// NativeTransferHandler handles `__CronosNativeTransfer` log
+type NativeTransferHandler struct {
+	bankKeeper   types.BankKeeper
+	cronosKeeper Keeper
+}
+
+func NewNativeTransferHandler(bankKeeper types.BankKeeper, cronosKeeper Keeper) *NativeTransferHandler {
+	return &NativeTransferHandler{
+		bankKeeper:   bankKeeper,
+		cronosKeeper: cronosKeeper,
+	}
+}
+
+func (h NativeTransferHandler) EventID() common.Hash {
+	return NativeTransferEvent.ID
+}
+
+func (h NativeTransferHandler) Handle(ctx sdk.Context, contract common.Address, data []byte) error {
+	unpacked, err := NativeTransferEvent.Inputs.Unpack(data)
+	if err != nil {
+		// log and ignore
+		h.cronosKeeper.Logger(ctx).Error("log signature matches but failed to decode", "error", err)
+		return nil
+	}
+
+	denom, found := h.cronosKeeper.GetDenomByContract(ctx, contract)
+	if !found {
+		return errors.New("contract is not connected to native token")
```

Comment:
I am thinking if we could wrap add those errors in the ethereum log, it could be nice

Reply:
can you add the contract address in the error?

Reply:
> I am thinking if we could wrap add those errors in the ethereum log, it could be nice

I believe that’s the case, the error returned by hook is saved in the return value of the EVM call.
But I need to double check on the rpc response.

Reply:
https://github.com/tharsis/ethermint/blob/main/x/evm/keeper/state_transition.go#L198

Hmm, the detailed error won't appear in the response, the RPC user only sees a fixed error string.
But it do get logged in a logfile as an error.

Reply:
okay I created a new issue for that #61